srk/src/polyhedron.ml

```diff
@@ -510,40 +636,281 @@ let equal p q =
   let dim = List.length cs in
   let csM = Linear.QQVectorSpace.matrix_of cs in
   if Linear.QQVectorSpace.equal cs (constraint_space q) then
-    let p0 = apron0_of man dim (invertible_image csM p) in
-    let q0 = apron0_of man dim (invertible_image csM q) in
-    Apron.Abstract0.is_eq man p0 q0
+    let p0 = nnc_dd_of ~man dim (invertible_image csM p) in
+    let q0 = nnc_dd_of ~man dim (invertible_image csM q) in
+    DD.equal p0 q0
   else
     false
 
-let of_generators dim generators =
-  let open Apron in
-  let man = Polka.manager_alloc_loose () in
-  let (vertices, rays) =
-    BatEnum.fold Generator0.(fun (vertices, rays) (kind, vec) ->
-        match kind with
-        | `Line -> (vertices, (make (lexpr_of_vec vec) LINE)::rays)
-        | `Ray -> (vertices, (make (lexpr_of_vec vec) RAY)::rays)
-        | `Vertex -> (vec::vertices, rays))
-      ([], [])
-      generators
+let valid_cone polyhedron =
+  let dim = max_constrained_dim polyhedron in
+  let (lines, rays) =
+    P.fold (fun (p, v) (lines, rays) ->
+        match p with
+        | `Zero -> (v::lines, rays)
+        | `Nonneg | `Pos -> (lines, v::rays))
+      polyhedron
+      ([], [V.of_term QQ.one Linear.const_dim])
   in
-  let polytope_of_point point =
-    Abstract0.of_lincons_array man 0 dim
-      (Array.init dim (fun i ->
-           (* x_i = point_i *)
-           let lexpr =
-             [(QQ.of_int (-1), i);
-              (V.coeff i point, Linear.const_dim)]
-             |> V.of_list
-             |> lexpr_of_vec
-           in
-           Lincons0.make lexpr Lincons0.EQ))
+  Cone.make ~lines ~rays dim
+
+let implies polyhedron (p, v) =
+  let c = valid_cone polyhedron in
+  match p with
+  | `Zero -> Cone.mem v c && Cone.mem (V.negate v) c
+  | `Nonneg -> Cone.mem v c
+  | `Pos -> invalid_arg "Polyhedron.implies does not currently support strict \
+                         inequalities"
+
+module NormalizCone = struct
+
+  open Normalizffi
+
+  let pp_vectors = Format.pp_print_list ~pp_sep:Format.pp_print_cut
+                     Linear.QQVector.pp
+
+  (* Rescale vector such that the selected coefficients are integral and
+     relatively prime *)
+  let normalize v =
+    Linear.QQVector.scalar_mul (QQ.inverse (Linear.QQVector.gcd_entries v)) v
+
+  module D = Linear.MakeDenseConversion(SrkUtil.Int)(V)
+
+  let densify ctx v =
+    let array = Array.make (D.dim ctx) (ZZ.mpz_of ZZ.zero) in
+    BatEnum.iter
+      (fun (a, i) ->
+         array.(D.int_of_dim ctx i) <- ZZ.mpz_of (Option.get (QQ.to_zz a)))
+      (V.enum v);
+    Array.to_list array
+
+  let sparsify ctx v =
+    BatList.fold_lefti (fun vec i a ->
+        V.set (D.dim_of_int ctx i) (QQ.of_zz (ZZ.of_mpz a)) vec)
+      V.zero
+      v
+
+  let normaliz_cone_by_constraints polyhedron =
+    let normalized_constraints =
+      enum_constraints polyhedron
+      /@ (fun (p, v) -> (p, normalize v))
+      |> (* add 1 >= 0 explicitly to make Normaliz cone pointed and
```

Comment:
This was a really bad comment :$ 

Normalizffi constructs the polyhedron by taking in homogeneous input and assuming that the first coordinate x0 corresponds to the constant dimension; the polyhedron is obtained from the cone by dehomogenizing the first coordinate, i.e., intersecting the cone with x0 = 1.

Sticking in 1 >= 0 here ensures that the constant dimension appears among the constraints, in case another dimension is mistakenly assumed as the constant dimension. 1 >= 0 is really presented as x0 >= 0 to Normalizffi.